cmd/dagger/cmd/do.go

```diff
@@ -21,27 +22,88 @@ import (
 )
 
 var doCmd = &cobra.Command{
-	Use:   "do [OPTIONS] ACTION [SUBACTION...]",
-	Short: "Execute a dagger action.",
+	Use: "do ACTION [SUBACTION...]",
+	// Short: "Execute a dagger action.",
 	PreRun: func(cmd *cobra.Command, args []string) {
 		// Fix Viper bug for duplicate flags:
 		// https://github.com/spf13/viper/issues/233
 		if err := viper.BindPFlags(cmd.Flags()); err != nil {
 			panic(err)
 		}
 	},
+	// Don't fail on unknown flags for the first parse
+	FParseErrWhitelist: cobra.FParseErrWhitelist{
+		UnknownFlags: true,
+	},
+	// We're going to take care of flag parsing ourselves
+	DisableFlagParsing: true,
 	Run: func(cmd *cobra.Command, args []string) {
-		if len(args) < 1 {
-			doHelpCmd(cmd, nil)
-			return
-		}
-
 		var (
 			lg  = logger.New()
 			tty *logger.TTYOutput
 			err error
 		)
 
+		cmd.Flags().Parse(args)
+
+		targetPath := getTargetPath(cmd.Flags().Args())
+
+		daggerPlan, err := loadPlan(viper.GetString("plan"))
+		if err != nil && !viper.GetBool("help") {
+			err = fmt.Errorf("failed to load plan: %w", err)
+			doHelpCmd(cmd, nil, nil, nil, targetPath, []string{err.Error()})
+			os.Exit(1)
+		} else if err != nil && viper.GetBool("help") {
+			doHelpCmd(cmd, nil, nil, nil, targetPath, nil)
+			os.Exit(0)
+		}
+
+		action := daggerPlan.Action().FindByPath(targetPath)
+
+		if action == nil {
```

Comment:
If I get this right, this finds the closest action correct?

Wouldn't it be confusing if you do a `dagger do test this does not exist` to get the help for `test`? Maybe not.

In any case, to make this code simpler, perhaps we could move this in `Action().FindClosest(targetPath)` or something?

Reply:
My thinking was that it would be helpful to know the available actions from the first valid action. Happy to change that up, if we prefer a different DX. Not attached to this.